src/hash/blake3.rs

```diff
@@ -0,0 +1,196 @@
+use wasm_bindgen::prelude::*;
+
+/// Generate a 24-byte nonce from input material using BLAKE3.
+/// - `nonce_material`: Raw bytes to derive the nonce from
+/// Returns 24 bytes suitable for use as a nonce in cryptographic operations.
+/// This function is deterministic - the same input will produce the same nonce.
+#[wasm_bindgen]
+pub fn generate_nonce(nonce_material: &[u8]) -> Vec<u8> {
+    let mut hasher = blake3::Hasher::new();
+    hasher.update(nonce_material);
+    hasher.finalize().as_bytes()[..24].to_vec()
+}
+
+/// Hash data once using BLAKE3.
+/// - `data`: Raw bytes to hash
+/// Returns 32 bytes of hash output.
+/// This is the simplest way to compute a BLAKE3 hash of a single piece of data.
+#[wasm_bindgen]
+pub fn blake3_hash_once(data: &[u8]) -> Vec<u8> {
+    let mut hasher = blake3::Hasher::new();
+    hasher.update(data);
+    hasher.finalize().as_bytes().to_vec()
+}
+
+/// Hash data once using BLAKE3 with a context prefix.
+/// - `data`: Raw bytes to hash
+/// - `context`: Context bytes to prefix to the data
+/// Returns 32 bytes of hash output.
+/// This is useful for domain separation - the same data hashed with different contexts will produce different outputs.
+#[wasm_bindgen]
+pub fn blake3_hash_once_with_context(data: &[u8], context: &[u8]) -> Vec<u8> {
+    let mut hasher = blake3::Hasher::new();
+    hasher.update(context);
+    hasher.update(data);
+    hasher.finalize().as_bytes().to_vec()
+}
+
+/// Get an empty BLAKE3 state for incremental hashing.
+/// Returns an empty vector representing the initial state.
+/// Use this to start an incremental hashing operation.
+#[wasm_bindgen]
+pub fn blake3_empty_state() -> Vec<u8> {
+    Vec::new()
```

Comment:
This is the only really important comment: Instead of appending data into a growing Vec and then hashing that in the end, please may you make the blake3 "state" a `blake3::Hasher`?

In `update_state` you should be able to do `hasher.update(new_data)` and in `digest_for_state` you just do `finalize()`

The tricky bit is probably passing the hasher back to JS - I think you should be able to wrap it in a wasm_bindgen exported struct with the hasher as a private field like so:

```rs
#[wasm_bindgen]
struct Blake3Hasher(blake3::Hasher);
```

That way you just pass an opaque handle back to JS.

The update function should probably take a `&mut Blake3Hasher` and digest just the owned `Blake3Hasher`

Reply:
Done, just finalising the changes on jazz's side